### PR TITLE
Revert conversion to Lite member during Split-Brain healing

### PR DESCRIPTION
**Pull request by ahmetmircik**
ee counterpart: https://github.com/hazelcast/hazelcast-enterprise/pull/2021

- reverts conversion to lite member
- disposes stores upon merge
- unifies hd merge code with heap merge codes, ee side merge code will be removed

closes https://github.com/hazelcast/hazelcast-enterprise/issues/1933
closes https://github.com/hazelcast/hazelcast/issues/12405

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/ClusterMergeTask.java [hunk 1]
```diff
@@ -106,19 +100,6 @@ private void disposeTasks(Collection<Runnable>... tasks) {
         }
     }
 
-    private void tryToPromoteLocalLiteMember() {
-        if (wasLiteMember) {
```
Comment (mdogan) [inline hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/ClusterMergeTask.java]: `wasLiteMember` field can be deleted too.

Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/ClusterMergeTask.java]: done

File: hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractCacheService.java [hunk 2]
```diff
@@ -228,7 +227,7 @@ public DistributedObject createDistributedObject(String cacheNameWithPrefix) {
                     cacheConfig.setManagerPrefix(HazelcastCacheManager.CACHE_MANAGER_PREFIX);
                 }
 
-                CacheMergePolicyProvider mergePolicyProvider = splitBrainHandlerService.getMergePolicyProvider();
+                CacheMergePolicyProvider mergePolicyProvider = new CacheMergePolicyProvider(nodeEngine);
```
Comment (Donnerbart) [inline hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractCacheService.java]: We should have a single `CacheMergePolicyProvider` instance per HZ node, which is used for the config check and the split-brain merging. The provider class caches the created merge policy instances, so with a single instance we know for sure that the merge policy instance, which survived the config check is used in the split-brain process later. It also saves us a bunch of `Class.for()` calls and garbage.

That's why the legacy merge policy providers and all other data structures use the single instance of `SplitBrainMergeProvider` from the `NodeEngine`.

Same applies for `IMap` and `ReplicatedMap`.

Comment (Donnerbart) [reply hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractCacheService.java]: So basically the services or map service context should store the created merge policy provider and make it accessible to the merge process.

Comment (Donnerbart) [reply hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractCacheService.java]: Maybe we can define a `LegacySplitBrainMergePolicyProvider` interface with a `Object getMergePolicy(String mergePolicyName)` method. Then we can store a single instance in the `AbstractSplitBrainHandlerService` and make it easily accessible for the config checks and merge runnables.

Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractCacheService.java]: changed it to one-instance-per-service.

**Comment by ahmetmircik** [comment]
run-lab-run

**Comment by ahmetmircik** [comment]
run-lab-run

File: hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/RecordStore.java [hunk 3]
```diff
@@ -381,6 +381,12 @@
 
     void destroy();
 
+    /**
+     * Like {@link #destroy()} but does not touch state on other services
+     * like lock service or event journal service.
+     */
+    void destroyInternals();
```
Comment (mmedenjak) [inline hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/RecordStore.java]: It's getting a bit cofusing - we have `destroy` (destroys the internals and the connected services), `destroyInternals` (destroys only the map internals), `reset` (I'm not sure how this one fits in, it also resets some data but it keeps some other like indexes and connected services like locks and journal) and `clearPartition` (just clears everything, also connected services).
I don't see a simple way of improving this. Maybe a task for a separate PR.

**Comment by ahmetmircik** [comment]
run-lab-run

File: hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/RecordStore.java (hunk 3 above)
Comment (tkountis) [reply hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/RecordStore.java]: Couldn't agree more :100:

Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/RecordStore.java]: This is needed for this PR to pass the tests but filed an improvement issue, this problem is a bit tricky. https://github.com/hazelcast/hazelcast/issues/12779

File: hazelcast/src/main/java/com/hazelcast/spi/impl/merge/AbstractSplitBrainHandlerService.java [hunk 4]
```diff
@@ -37,122 +39,94 @@
...
-                    Store store = iterator.next();
+        ConcurrentLinkedQueue<Store> mergingStores = new ConcurrentLinkedQueue<Store>();
+        ConcurrentLinkedQueue<Store> notMergingStores = new ConcurrentLinkedQueue<Store>();
 
-                    if (partitionOwner) {
-                        String dataStructureName = getDataStructureName(store);
-                        Object mergePolicy = getMergePolicy(dataStructureName);
+        collectStores(mergingStores, notMergingStores);
 
-                        if (!isDiscardPolicy(mergePolicy)) {
```
[the remark this answers is not in the document]
Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/spi/impl/merge/AbstractSplitBrainHandlerService.java]: put this checks to split-brain-handler-service.

File: hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractCacheService.java [hunk 5]
```diff
@@ -142,15 +143,20 @@ public final void init(NodeEngine nodeEngine, Properties properties) {
...
         this.eventJournal = new RingbufferCacheEventJournalImpl(nodeEngine);
+        this.mergePolicyProvider = new CacheMergePolicyProvider(nodeEngine);
+
         postInit(nodeEngine, properties);
     }
 
-    // this method is overridden on ee
-    protected CacheSplitBrainHandlerService newSplitBrainHandlerService(NodeEngine nodeEngine) {
-        return new CacheSplitBrainHandlerService(nodeEngine, configs, segments);
+    public CacheMergePolicyProvider getMergePolicyProvider() {
```
[the remark this answers is not in the document]
Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractCacheService.java]: kept this as is, it is being used by different packages on ee side

**Comment by pveentjer** [comment]
@ahmetmircik can you rebase?

File: hazelcast/src/main/java/com/hazelcast/spi/impl/merge/BaseSplitBrainHandlerService.java [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+                asyncDestroyStores(storesToDestroy, partitionId);
+
+            } finally {
+                latch.countDown();
+            }
+        }
+    }
+
+    void asyncDestroyStores(final Collection<Store> stores, final int partitionID) {
+        operationExecutor.execute(new PartitionSpecificRunnable() {
```
Comment (mmedenjak) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/merge/BaseSplitBrainHandlerService.java]: Can the partition be migrating? Do we need to retry? See - https://github.com/hazelcast/hazelcast/pull/12691#discussion_r178733516

Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/spi/impl/merge/BaseSplitBrainHandlerService.java]: No need to retry, `PartitionSpecificRunnable ` can independently run on local without doing any checks like `is-migration-in-progress`.

**Pull request merged**